File: algorithms/medium/incremental_memory_leak.go
package main

import "fmt"

func memLeak(memory1 int, memory2 int) []int {
    i := 0
    for true {
        i++
        if memory1 < i && memory2 < i {
            return []int{i, memory1, memory2}
        }
        if memory1 >= memory2 {
            memory1 -= i
        } else {
            memory2 -= i
        }
    }
    return []int{}
}

func main() {
    data := [][]int{
        []int{2, 2},
        []int{8, 11},
        []int{1234567890, 2147483647},
        []int{0, 0},
        []int{0, 10},
    }
    for _, arr := range data {
        memory1, memory2 := arr[0], arr[1]
        fmt.Println(memory1, memory2)
        r := memLeak(memory1, memory2)
        fmt.Println(r)
        fmt.Println("===================")
    }
}

File: algorithms/medium/incremental_memory_leak.py
from typing import List

class Solution:
    def memLeak(self, memory1: int, memory2: int) -> List[int]:
        i = 0
        while True:
            i += 1
            if memory1 < i and memory2 < i:
                return [i, memory1, memory2]
            if memory1 >= memory2:
                memory1 -= i
            else:
                memory2 -= i

# Main section
for memory1, memory2 in [
                           (2, 2),
                           (8, 11),
                           (1234567890, 2147483647),
                           (0, 0),
                           (0, 10),
                        ]:
    print(f'memory1, memory2 = {memory1}, {memory2}')
    sol = Solution()
    r = sol.memLeak(memory1, memory2)
    print(f'r = {r}')
    print('===============================')


